controller/src/main/java/io/pravega/controller/server/rpc/auth/PravegaInterceptor.java

```diff
@@ -0,0 +1,117 @@
+import io.grpc.ServerInterceptor;
+import io.grpc.Status;
+import io.jsonwebtoken.Jwts;
+import io.jsonwebtoken.SignatureAlgorithm;
+import io.pravega.auth.PravegaAuthHandler;
+import java.util.HashMap;
+import java.util.Map;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+
+import static io.pravega.auth.PravegaAuthHandler.PravegaAccessControlEnum.READ_UPDATE;
+
+@Slf4j
+public class PravegaInterceptor implements ServerInterceptor {
+    private static final boolean AUTH_ENABLED = true;
+    private static final String AUTH_CONTEXT = "PravegaContext";
+    private static final String INTERCEPTOR_CONTEXT = "InterceptorContext";
+    private static final Context.Key<Map<String, String>> AUTH_CONTEXT_PARAMS = Context.key(AUTH_CONTEXT);
+    private static final Context.Key<PravegaInterceptor> INTERCEPTOR_OBJECT = Context.key(INTERCEPTOR_CONTEXT);
+
+    private final PravegaAuthHandler handler;
+    @Getter
+    private String delegationToken;
+
+    PravegaInterceptor(PravegaAuthHandler handler) {
+        Preconditions.checkNotNull(handler, "handler can not be null");
+        this.handler = handler;
+    }
+
+    @Override
+    public <ReqT, RespT> ServerCall.Listener<ReqT> interceptCall(ServerCall<ReqT, RespT> call, Metadata headers, ServerCallHandler<ReqT, RespT> next) {
+
+        Map<String, String> paramMap = new HashMap<>();
+        headers.keys().stream().forEach(key -> {
+            try {
+                paramMap.put(key,
+                        headers.get(Metadata.Key.of(key, Metadata.ASCII_STRING_MARSHALLER)));
+            } catch (IllegalArgumentException e) {
+                log.warn("Error while marshalling some of the headers", e);
+            }
+        });
+        String method = paramMap.get("method");
+        Context context = Context.current();
+        if (!Strings.isNullOrEmpty(method)) {
+            if (method.equals(handler.getHandlerName())) {
+
+                if (!handler.authenticate(paramMap)) {
+                    call.close(Status.fromCode(Status.Code.UNAUTHENTICATED), headers);
+                    return null;
+                }
+                context = context.withValue(AUTH_CONTEXT_PARAMS, paramMap);
+                context = context.withValue(INTERCEPTOR_OBJECT, this);
+            }
+        }
+        return Contexts.interceptCall(context, call, headers, next);
+    }
+
+    public static PravegaInterceptor getCurrentInterceptor() {
+        return INTERCEPTOR_OBJECT.get();
+
+    }
+
+    public PravegaAuthHandler.PravegaAccessControlEnum authorize(String resource) {
+        return this.handler.authorize(resource, AUTH_CONTEXT_PARAMS.get());
+    }
+
+    public static String retrieveDelegationToken(String tokenSigningKey) {
+        PravegaInterceptor interceptor;
+        if ((interceptor = getCurrentInterceptor()) != null) {
+            return interceptor.getDelegationToken();
+        } else {
+            Map<String, Object> claims = new HashMap();
```

Comment:
Shouldn't this return null? Not having currentInterceptior set causing everything to be unconditionally authorized sounds like a nasty bug waiting to happen.

Reply:
This is the default case. Where there is no interceptor registered, it means that auth is disabled. Otherwise it will not be set.